requestStorageAccessForOrigin.bs

```diff
@@ -0,0 +1,228 @@
+1. Let |parsedURL| be the the result of running the [=URL parser=] on |requestedOrigin|.
+1. If |parsedURL| is failure, [=reject=] |p| with a "{{TypeError}}" {{DOMException}} and return |p|.
+1. Let |origin| be |parsedURL|'s [=/origin=].
+1. If |origin| is an [=opaque origin=], [=reject=] |p| with an "{{InvalidStateError}}" {{DOMException}} and return |p|.
+1. If |doc|'s [=Document/origin=] is [=same origin=] with |origin|, [=resolve=] and return |p|.
+1. Let |descriptor| be a newly created {{TopLevelStorageAccessPermissionDescriptor}} with {{PermissionDescriptor/name}} set to "<a permission><code>top-level-storage-access</code></a>" and with {{TopLevelStorageAccessPermissionDescriptor/requestedOrigin}} set to |origin|.
+1. Run these steps [=in parallel=]:
+    1. Let |top-level site| be the result of [=obtain a site|obtaining a site=] from |doc|'s [=relevant settings object=]'s [=environment settings object/origin=].
+    1. [=Determine the top-level storage access policy=] with |descriptor|, |doc|, |top-level site|, and |p|.
+1. Return |p|.
+
+</div>
+
+<h3 id="ua-policies">User Agent top-level storage access policies</h3>
+
+<div algorithm>
+To <dfn type="abstract-op">determine if a request has top-level storage access</dfn> with [=request=] |request|, run these steps:
+
+1. Let |settings| be |request|'s [=request/client=]'s [=relevant global object=]'s [=relevant settings object=].
+1. Let |embedded origin| be |request|'s [=request/url=]'s [=/origin=].
+1. Let |descriptor| be a newly created {{TopLevelStorageAccessPermissionDescriptor}} with {{PermissionDescriptor/name}} set to "<a permission><code>top-level-storage-access</code></a>" and with {{TopLevelStorageAccessPermissionDescriptor/requestedOrigin}} set to |embedded origin|.
+1. Let |existing state| be |descriptor|'s [=permission state=] with |settings|.
+1. If |existing state| is [=permission/granted=], return true.
+1. Return false.
+
+</div>
+
+<div algorithm>
+To <dfn type="abstract-op">determine the top-level storage access policy</dfn> for {{TopLevelStorageAccessPermissionDescriptor}} |descriptor|, with {{Document}} |doc|, [=site=] |top-level site|, and {{Promise}} |p|, run these steps:
+1. Let |settings| be |doc|'s [=relevant settings object=].
+1. Let |global| be |doc|'s [=relevant global object=].
+1. Let |existing state| be |descriptor|'s [=permission state=] with |settings|.
+1. If |existing state| is [=permission/granted=]:
+    1. [=Queue a global task=] on the [=permission task source=] given |global| to [=resolve=] |p|.
+    1. Return.
+1. If |existing state| is [=permission/denied=]:
+    1. [=Queue a global task=] on the [=permission task source=] given |global| to [=reject=] |p| with a "{{NotAllowedError}}" {{DOMException}}.
+    1. Return.
+1. Assert that |doc|'s [=node navigable=] is a [=traversable navigable=].
+1. If this algorithm was invoked when |doc|'s {{Window}} object did not have [=transient activation=]:
```

Comment:
This is a bit handwavy as we're doing some action-at-a-distance, looking backwards to state that lived on a different thread. I think it's best to pull the state of the user activation out early on the main thread and plumb it through